VV refactor (In which mso adds a bunch of shit to VV while still removing a net 200 lines of code)

MrsTonedOne opened the pull request:
VV refactor:

LIST VV:
![image](https://cloud.githubusercontent.com/assets/7069733/20640190/6a60e8a6-b38c-11e6-918a-cf9d0c1e421d.png)

Modular ALL THE THINGS

VV's dropdown is a datum proc for easy adding shit to it.

Calculate on display vars (just override vv_get_var())

on_varedit has been replaced with vv_edit_var, this proc does the actual editing, so objects can decide some of it's vars aren't editable by just not calling parent in here. Removes snowflake from vv.

350+ lines removed from modify variables and mass modify without removing functionality

Copy/Paste removed from mass modify. Editing it is much much easier.

Fixed a million bugs with auto-detecting type.

Expanded type selection for editing variables (You can now choose between multiple line and normal input for text, as well as pick clients, datums, hand typed types, etc)

Unified input. Everything uses the same 2 procs for getting the variable from the user, so now everything supports all the same shit, and expansion is easy. (This was added to call proc as well)

Mass modify now supports clients and datums, as well as mass modifying to a new atom or new datum or new list (and an option to choose between one or many)

Pressing f5 in vv now actually reloads the page.

The search/filter value is kept between loads for the same object

Picking a reference was made nicer, filters by selectable type.

The number of strings created in the string table by VVing has been massively reduced.

Fixes #19074

@Shadowlight213 @Iamgoofball @MrPerson @tgstation/commit-access @Aranclanos @PJB3005 @d3athrow @N3X15 @clusterfack @Fox-McCloud

MrsTonedOne committed:
VV refactor

Saegrimr commented:
>The search/filter value is kept between loads for the same object

YES

PJB3005 commented:
\> He copy pasted my name from MoMMI relaying my messages to IRC it blew the mention.

👌

Supermichael777 commented:
BasedMSO

the coderhost we do not deserve

RemieRichards commented on code/modules/admin/verbs/modifyvariables.dm:
```diff
@@ -94,401 +206,205 @@ var/list/VVckey_edit = list("key", "ckey")
-	if(!class)
-		return
+	if(default == VV_NUM)
+		var/dir_text = ""
```
could probably do with a proc for this.
dir2dirtext or something (it might actually exist already)

RemieRichards commented on code/modules/mob/mob.dm:
```diff
@@ -852,31 +852,40 @@ var/next_mob_id = 0
+			if((stat == 2) && (var_value < 2))//Bringing the dead back to life
+				dead_mob_list -= src
+				living_mob_list += src
+			if((stat < 2) && (var_value == 2))//Kill he
```
```DEAD```, etc.

MrStonedOne replied:
where are the stat defines even at so i can fix this to be proper. Its shit code i evicted from https://github.com/tgstation/tgstation/pull/21778/files#diff-8d921f8d3059a9878269cbe94351928bL632 so I didn't even pay attention to it's contents.

RemieRichards replied:
the thumbs up means it's resolved JJ.

RemieRichards commented on code/modules/mob/mob.dm:
```diff
@@ -852,31 +852,40 @@ var/next_mob_id = 0
+	switch(var_name)
 		if("weakened")
-			SetWeakened(weakened)
+			SetWeakened(var_value)
```
for atom colours you chose to use ```.color``` here, from the ```. = ..()``` part of the VV edit.
but here you choose to use ```var_value```, any particular reason? the inconsistency just seems odd.

MrStonedOne replied:
I started this 2 months ago and it sat mostly unfinished for 2 months, the number of conflicts i had to fix was incredible that i just kept a bunch of the code when fixing them.

This is some of the code i did 2 months ago, color how ever was a conflict

RemieRichards reviewed (commented):
Goooooood Stuuuuuff.
Just some minor quibbles.

MrStonedOne commented on code/controllers/subsystem.dm:
```diff
@@ -157,8 +157,12 @@
+		if ("can_fire")
+			if (var_value)
+				next_fire = world.time + wait
+		if ("queued_priority") //editing this breaks things.
```
if it returns 0, the calling call will tell the user

MrStonedOne commented on code/datums/datumvars.dm:
```diff
@@ -1,43 +1,53 @@
-			body += debug_variable("icon", new/icon(A.icon, A.icon_state, A.dir), 0)
-		#endif
 
+	var/islist = FALSE
```
very good idea, I love it!

MrsTonedOne committed:
Improvements and fixes

MrsTonedOne committed:
>trusting admins

MrsTonedOne committed:
Makes the proccall admin verbs use the vv input grabber procs

MrsTonedOne committed:
Fixes modifing null values with (e) not working.

MrsTonedOne committed:
Some shitcode tried to escape. @Shadowlight123 caught it.

MrStonedOne commented:
I found more shitcode to remove.

Jordie0608 commented:
Don't look too much or you'll end up removing the entire codebase.

MrsTonedOne committed:
Adds some dropdown options to lists
I was gonna make it so you could mark lists, but too much code assumed they had a .type and adding the proper istypes to them was too bloaty.

MrStonedOne commented:
I was going to refactor the vv topic if else tree, but that was gonna take too long and i want to just be done with this pr.

I'm testing my more recent changes now, then it should be ready to go

MrsTonedOne committed:
Fixes some bugs/runtimes. Makes vv var formating use less strings.

optimumtact commented:
@tgstation/commit-access

Iamgoofball commented:
god bless

AnturK commented on code/datums/datumvars.dm:
```diff
@@ -387,53 +422,47 @@ body
-			html += "</ul>"
+		var/list/items = list()
+
+		if (L.len > 0 && !(name == "underlays" || name == "overlays" || L.len > 500))
```
I'd make it even lower, really can't think of where i wouldn't want list of size > 100 open in another window.

AnturK commented on code/modules/admin/verbs/modifyvariables.dm:
```diff
@@ -1,494 +1,520 @@
-		if("num")
-			var_value = input("Enter new number:","Num") as num
+		if (VV_ICON)
+			.["value"] = input("Pick icon:", "Icon") as null|icon
```
Would be nice if icon_state was populated from icon_states() but that can wait for another PR

AnturK commented on code/datums/datumvars.dm:
```diff
@@ -1,220 +1,103 @@
-			</font>
-
-
+			atomsnowflake += {"
```
Could be per type so we can have fancy displays for common objects, then again, don't expect anyone to waste time on that so probably pointless.

AnturK reviewed (approved):
Looks good, can't see anything obviously wrong. (this is typo breeding ground kind of code though)

lordpidey commented on code/datums/datumvars.dm:
```diff
@@ -223,156 +98,291 @@ var/global/list/internal_byond_list_vars = list("contents" = TRUE, "verbs" = TRU
+	if (!islist)
+		for (var/V in D.vars)
+			names += V
+	sleep(1)//For some reason, without this sleep, VVing will cause client to disconnect on certain objects.
```
This might actually have been fixed with byond 511.
